casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -303,42 +300,46 @@ sealed abstract class MultiParentCasperInstances {
+          case InvalidUnslashableBlock =>
+            handleInvalidBlockEffect(status, block)
           case InvalidBlockNumber =>
-            handleInvalidBlockEffect(block)
+            handleInvalidBlockEffect(status, block)
           case InvalidParents =>
-            handleInvalidBlockEffect(block)
+            handleInvalidBlockEffect(status, block)
+          case JustificationRegression =>
+            handleInvalidBlockEffect(status, block)
           case InvalidSequenceNumber =>
-            // We need to handle invalid sequence numbers separately as the invalid block tracker
-            // assumes the sequence numbers inserted are valid.
-            throw new Error("InvalidSequenceNumber unimplemented")
-          case InvalidUnslashableBlock =>
-            for {
-              _ <- Log[F].info(
-                    s"CASPER: Did not add invalid block ${PrettyPrinter.buildString(block.blockHash)}")
-            } yield ()
+            handleInvalidBlockEffect(status, block)
           case _ => throw new Error("Should never reach")
         }
 
-      private def handleInvalidBlockEffect(block: BlockMessage): F[Unit] =
+      private def handleInvalidBlockEffect(status: BlockStatus, block: BlockMessage): F[Unit] =
         for {
-          _ <- Log[F].info(
-                s"CASPER: Did not add invalid block ${PrettyPrinter.buildString(block.blockHash)}")
-          _ <- Capture[F].capture {
-                val invalidBlockRecord =
-                  invalidBlockTracker.getOrElse(block.sender, InvalidBlockRecord())
-                val updatedInvalidBlockRecord = invalidBlockRecord.invalidBlockSeqNum match {
-                  case Some(_) => invalidBlockRecord
-                  case None =>
-                    InvalidBlockRecord(Some(block.seqNum),
-                                       invalidBlockRecord.invalidBlockDiscoveryStatuses)
-                }
-                invalidBlockTracker.update(block.sender, updatedInvalidBlockRecord)
-              }
+          _ <- Log[F].info(s"CASPER: Did not add invalid block ${PrettyPrinter.buildString(
+                block.blockHash)} for ${status.toString}.")
+          // TODO: Slash block.blockHash for status except InvalidUnslashableBlock
+          blocksToSlash = allChildren[BlockHash](awaitingJustificationToChild, block.blockHash) - block.blockHash
+          _ <- Log[F].warn(s"""CASPER: About to slash the following blocks ${blocksToSlash
+                .map(PrettyPrinter.buildString)
+                .mkString("", ",", "")} for neglecting an invalid block""")
+          // TODO: Slash blocksToSlash for neglecting an invalid block
+          _ <- Capture[F].capture { blocksToSlash.map(awaitingJustificationToChild.remove) }
         } yield ()
 
+      private def allChildren[A](map: mutable.MultiMap[A, A], element: A): Set[A] = {
```

Comment:
I think @goral09 is right about refactoring this using `DagOperations.bfTraverse`
```scala
private def allChildren[A](map: mutable.MultiMap[A, A], element: A): Set[A] = 
  DagOperations.bfTraverse[A](Some(element)){
    (el: A) => map.getOrElse(el, Set.empty[A])
  }.toSet
```